### DISCUSSION TITLE
Open a random link from a markdown file

### DISCUSSION THREAD

**Discussion by horacioh** [johnlindquist/kit#618] [Share]
I have a list of all the links I want to read in a markdown file. and I build this script to open a random link to start reading!

```js
/** @type {import("@johnlindquist/kit")} */

// Menu: Read Random link from a mardown file
// Description: Open a random link from a markdown file
// Author: Horacio Herrera
// Twitter: @hhg2288

/** @type {typeof import("unified")} */
let { unified } = await npm("unified")
/** @type {typeof import("remark-parse").default} */
let remarkParse = await npm("remark-parse")
/** @type {typeof import("unist-util-visit")} */
let { visit } = await npm("unist-util-visit")

// get the markdown file you want to parse
let file = home(`workspace/braindump/inbox.md`)
let contents = await readFile(file, "utf-8")
let tree = unified().use(remarkParse).parse(contents)

// create an array of links
let links = []
visit(tree, "link", function linkVisitor(node) {
  links.push(node)
})

// open a random url from the generated list
let index = Math.floor(Math.random() * (links.length - 1))
exec(`open '${links[index].url}'`)

```

there are some things I want to modify to make it better and more "agnostic":
- [x] let de user set the file he/she wants to parse
- [ ] store the remark AST result in the db, and maybe update it everytime I update the file (which I do using another script)
- [ ] update the file and create another list of the links I "visited"

any feedback and suggestion is welcome!

**Comment by johnlindquist** [comment]
This is great! Are you using a script to generate this markdown file?

For the first option, 

```js
await selectFile("Pick a markdown file:")
```

 will open a file selection prompt.

For the third, `replace` might help, it's aliased to the  [https://www.npmjs.com/package/replace-in-file](https://www.npmjs.com/package/replace-in-file) lib.

**Reply by horacioh** [reply]
I'm using a script to add links to that file, I created it manually

Here's the script that updates the file:
```js
// Menu: Add Link to Inbox
// Description: Add the current browser focus link to the inbox file
// Author: Horacio Herrera
// Twitter: @hhg2288

let file = home(`workspace/braindump/inbox.md`) //"home" is the same as `~/whatever.md`

//returns a comma-delimited string: foo.com, Some title
let urlAndTitle = await applescript(
  String.raw`tell application "Brave Browser" to return {URL, title} of active tab of front window`
)

let [url, title] = urlAndTitle.split(/, (.+)/) //split on first comma
let link = `[${title}](${url})`

let contents = await readFile(file, "utf-8")

let insertedLink = contents.replace(
  /(# Inbox\n)/,
  `$1
- ${link}`
)

await writeFile(file, insertedLink)
```

to remove the link from the original list I was thinking of using unified to do it, and to add the link back to the "visited" list, maybe the same technique I use to add the link in the above script might work!

is there a way to persist the file I select using `await selectFile("Pick a markdown file:")`?

**Reply by johnlindquist** [reply]
Sure, `env` can take a function:

[Install select-file](https://scriptkit.com/api/new?name=select-file&url=https://gist.githubusercontent.com/johnlindquist/d14047e72abbb7f8631f618fc7abfce6/raw/60bcf44bb4d2437b9852989af564c2f8827d57ea/select-file.js")

```js
import "@johnlindquist/kit"

let filePath = await env("SELECTED_MARKDOWN_FILE", () =>
  selectFile("Pick a markdown file")
)

await div(filePath)
```
